### PR TITLE
DOC Explain how to get off CWP

### PR DESCRIPTION
**Pull request by maxime-rainville**
Parent issue: https://github.com/silverstripeltd/product-issues/issues/468

### CODE REVIEW

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+The _tilde constraint_ is a suitable default and is the constraint style that ships in the `silverstripe/installer`.
+
+Read [Next Significant Release Operators](https://getcomposer.org/doc/articles/versions.md#next-significant-release-operators) in the composer documentation for more information version constraint operators.
+
+### What if I get a conflict when running composer update?
+
+### What happens if I do not migrate off CWP?
+
+### Is there a way to create a new project that would require all the modules a CWP project would have?
```
Comment (maxime-rainville) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:44]: I added a couple questions that I think might spring to mind here, but I would rather have some validations that we think those are usefull questions to answer before I spent too much time on them.

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+# Migrate your project off the Common Web Platform v2
+
+Until November 2021, Silverstripe Ltd maintained a specilasied version of the Silverstripe CMS project for the New Zealand Goverment and related New Zealand public agencies. This version was called _Common Web Platform_ (CWP for short).
+
+The CWP CMS release has been discontinued. CWP project who wish to continue to receive update most migrate their code base off the latest CWP release back to the regular Silverstripe CMS release.
+
+## Steps to migcate away from CWP
+
+Before you began this process you must identify which version of Silverstripe CMS Recipe you want to upgrade to. You can upgrade to [Silverstripe CMS Recipe 4.9.0](/changelogs/4.9.0/) or greater.
```
Comment (michalkleiner) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:14]:
```suggestion
Before you begin this process you must identify which version of Silverstripe CMS Recipe you want to upgrade to. You can upgrade to [Silverstripe CMS Recipe 4.9.0](/changelogs/4.9.0/) or greater.
```

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
+---
+title: Migrating off CWP
+summary: Migrate your project off the Common Web Platform v2
+---
+
+# Migrate your project off the Common Web Platform v2
+
+Until November 2021, Silverstripe Ltd maintained a specilasied version of the Silverstripe CMS project for the New Zealand Goverment and related New Zealand public agencies. This version was called _Common Web Platform_ (CWP for short).
```
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:8]:
```suggestion
Until November 2021, Silverstripe Ltd maintained a specialised version of the Silverstripe CMS project for the New Zealand Government and related New Zealand public-sector agencies. This version was called _Common Web Platform_ (CWP for short).
```

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+The CWP CMS release has been discontinued. CWP project who wish to continue to receive update most migrate their code base off the latest CWP release back to the regular Silverstripe CMS release.
```
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:10]:
```suggestion
The CWP CMS release has been discontinued. CWP projects who wish to continue to receive updates must migrate their code base off the latest CWP release back to the regular Silverstripe CMS release.
```

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+## Steps to migcate away from CWP
+
+Before you began this process you must identify which version of Silverstripe CMS Recipe you want to upgrade to. You can upgrade to [Silverstripe CMS Recipe 4.9.0](/changelogs/4.9.0/) or greater.
+
+[Review the changelog list](/changelogs/) to choose the Silverstripe CMS Recipe version you wish to upgrade to. Read the changelog carefully. It may contain additional release specific upgrade guidance beyond the steps in this guide.
+
```
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:17]: Do projects need to be using [silverstripe/recipe-plugin](https://github.com/silverstripe/recipe-plugin) to run the below commands? If so, there's probably a need for a 'prerequisites section?

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+5. Review all your composer requirements and update the constraints so the module versions documented in the changelog for you desired Silverstripe CMS Recipe release are installable.
+6. Run a `composer update` to get the lastest tag of each module.
+
+## Questions
+
+### What constraint should I use in my requirement version?
+
+That really depends on how comfortable your organisiation is with automatically applying updates to your Silverstripe CMS project.
+
+The three broad constraint categories you can use are, in descending order of restrictiveness:
```
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:31]: I'm wondering if using 'silverstripe/framework' is the best example here? As if a project follows the steps above, they won't have this in their composer.json, they should just have recipe-cms.

Perhaps refer to something like userforms, which was included in the cwp-installer.

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md (hunk 5 above)
Comment (maxime-rainville) [reply docs/en/03_Upgrading/03_Migrating_off_CWP.md:17]: `recipe-plugin` is a requirement of all recipes. You can't have a recipe without having it.

Comment (brynwhyman) [reply docs/en/03_Upgrading/03_Migrating_off_CWP.md:17]: Alright. Resolving.

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+The CWP CMS release has been discontinued. CWP projects who wish to continue to receive updates must migrate their code base off the latest CWP release back to the regular Silverstripe CMS release.
+
+## Steps to migcate away from CWP
+
+Before you begin this process you must identify which version of Silverstripe CMS Recipe you want to upgrade to. You can upgrade to [Silverstripe CMS Recipe 4.9.0](/changelogs/4.9.0/) or greater.
+
+[Review the changelog list](/changelogs/) to choose the Silverstripe CMS Recipe version you wish to upgrade to. Read the changelog carefully. It may contain additional release specific upgrade guidance beyond the steps in this guide.
+
+1. If your current composer file contains `cwp/cwp-installer`, run `composer update-recipe cwp/cwp-installer` to inline the recipe.
```
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:18]: I think these commands should be `composer require-recipe ...`, [as per the docs](https://github.com/silverstripe/recipe-plugin).

Comment (maxime-rainville) [reply docs/en/03_Upgrading/03_Migrating_off_CWP.md:18]: I went through the doc and ran some local test. 

I think `composer require-recipe` is meant to install and inline a recipe that you have not yet installed in your project. If you call `composer require-recipe` for a recipe that you have already installed, nothing bad will happen. It will just inline the recipe, if it has not been inline already.

I think `composer udate-recipe` is meant to inline a recipe that you have already installed in your project. If you call `composer update-recipe` for a recipe that you NOT already installed, an error will be thrown. Arguably, this is a more desirable behaviour because it avoids the scenario of someone blindly following the instruction in the guide and accidentally installing a recipe.

TLDR Both `composer require-recipe` and `composer update-recipe` could work for this guide and will lead to the same outcome in most cases. `composer update-recipe` will however throw a potentially useful error if you try to inline a recipe you didn't mean to install.

**Comment by maxime-rainville** [comment]
The last commit:
- change the name of the substitute recipes
- added a blurb about cleaning up some extra composer metadada
- added answers for all the questions

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+Composer normally tells you which packages are in conflict. You can try using the `--with-all-dependencies` flag and allow composer upgrade or downgrade other dependencies to resolve conflicts:
+```bash
+composer update --with-all-dependencies
+```
+
+If the `--with-all-dependencies` flag doesn't fix the problem, carefully read the composer error message and try to identify which constraints need to be updated to resolve to an installable set of dependencies. 
+
+### What happens if I do not migrate off the CWP CMS recipes?
+
+The CWP CMS recipes have been archived, but existing versions will remained installable for the indefinite future. No new versions of the CWP CMS recipes will be tagged. The CWP CMS recipes are locksteped to exact versions of various Silverstripe CMS modules. This means you will not be able to upgrade to any Silverstripe CMS Recipes beyond the 4.8 release.
```
Comment (maxime-rainville) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:59]: > The CWP CMS recipes have been archived

I'm using the past tense here, but this has not occurred yet.

**Review by emteknetnz — APPROVED** [review]

**Comment by emteknetnz** [comment]
Approved assuming that we want this in the open source framework docs and somewhere else

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md (hunk 7 above)
Comment (brynwhyman) [reply docs/en/03_Upgrading/03_Migrating_off_CWP.md:18]: Ah, thanks for clarifying. I got that wrong :)

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md (hunk 6 above)
Comment (maxime-rainville) [reply docs/en/03_Upgrading/03_Migrating_off_CWP.md:39]: I switched to userforms

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+## Steps to migrate away from CWP
+
+Before you begin this process you must identify which version of Silverstripe CMS Recipe you want to upgrade to. You can upgrade to [Silverstripe CMS Recipe 4.9.0](/changelogs/4.9.0/) or greater.
+
+[Review the changelog list](/changelogs/) to choose the Silverstripe CMS Recipe version you wish to upgrade to. Read the changelog carefully. It may contain additional release specific upgrade guidance beyond the steps in this guide.
+
+1. If your current composer file contains `cwp/cwp-installer`, run `composer update-recipe cwp/cwp-installer` to inline the recipe.
+2. If your current composer file contains `cwp/cwp-recipe-cms`, run `composer update-recipe cwp/cwp-recipe-cms` to inline the recipe.
+3. If your current composer file contains `cwp/cwp-recipe-core`, replace it with `silverstripe/recipe-ccl`.
+4. If your current composer file contains `cwp/cwp-recipe-search`, replace it with `silverstripe/recipe-solr-search`.
```
Comment (maxime-rainville) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:21]: Need to clarify what numbering version format this will use.

File: docs/en/03_Upgrading/03_Migrating_off_CWP.md (hunk 8 above)
Comment (brynwhyman) [inline docs/en/03_Upgrading/03_Migrating_off_CWP.md:59]:
```suggestion
The CWP CMS recipes have been archived, but existing versions will remain installable for the indefinite future. No new versions of the CWP CMS recipes will be tagged. The CWP CMS recipes are in lockstep with exact versions of various Silverstripe CMS modules. This means you will not be able to upgrade to any Silverstripe CMS Recipes beyond the 4.8 release.
```

**Comment by brynwhyman** [comment]
I support this going into the open-source docs. 

My feeling is that in the near term we should create a new section on the CMS Dev Docs that relates to recipes. Under this, we can get specific about each recipe and any relevant additional information. [The idea of focusing more on recipes has already been started here](https://github.com/silverstripe/silverstripe-framework/pull/10084).

The docs in this PR could then be linked from, or moved to, this new section under something along the lines of "NZ Public Sector Recipe' (or whatever we choose to call it), something like: `.../Module Recipes/NZ Public sector recipes/Migrating off CWP Guide`

File: docs/en/03_Upgrading/03_Upgrading/03_Migrating_off_CWP.md [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+1. If your current composer file contains `cwp/cwp-installer`, run `composer update-recipe cwp/cwp-installer` to inline the recipe.
+2. If your current composer file contains `cwp/cwp-recipe-cms`, run `composer update-recipe cwp/cwp-recipe-cms` to inline the recipe.
+3. If your current composer file contains `cwp/cwp-recipe-core`, replace it with `silverstripe/recipe-ccl`.
+4. If your current composer file contains `cwp/cwp-recipe-search`, replace it with `silverstripe/recipe-solr-search`.
+5. Review all your composer requirements and update the constraints so the module versions documented in the changelog for you desired Silverstripe CMS Recipe release are installable.
+6. Run a `composer update` to get the lastest tag of each module.
+
+The `silverstripe/recipe-solr-search` and `silverstripe/recipe-ccl` release versions follow the same numbering scheme as the recipes they superseded:
+- the last release version of `cwp/cwp-recipe-search` and `cwp/cwp-recipe-core` is 2.8.0
+- the first release version of `silverstripe/recipe-solr-search` and `silverstripe/recipe-ccl` is 2.9.0.
```
Comment (maxime-rainville) [inline docs/en/03_Upgrading/03_Upgrading/03_Migrating_off_CWP.md:27]: @brynwhyman  That's the only bit I added

**Review by brynwhyman — APPROVED** [review]
